src/storage/src/source/generator/ldbc.rs

```diff
@@ -0,0 +1,247 @@
+                        error!("could not find ldbc relation: {entity}");
+                        continue;
+                    };
+                    (output, desc)
+                }
+                Some(_) => continue,
+                None => continue,
+            };
+            let reader = Bytes::from(entry).reader();
+            let rdr = ReaderBuilder::new().delimiter(b'|').from_reader(reader);
+            self.rdr = Some(CsvFile {
+                rdr,
+                desc,
+                output: *output,
+            });
+        }
+    }
+
+    /// Reads the next record out of `self.rdr`. Returns `None` if `self.rdr` is out of records.
+    fn next_record(&mut self) -> Option<(usize, Row)> {
+        self.rdr
+            .as_mut()
+            .and_then(|csv| match csv.rdr.read_record(&mut self.record) {
+                Ok(true) => {
+                    let row = Row::pack(self.record.iter().zip(csv.desc.iter_types()).map(
+                        |(field, typ)| {
+                            let pgtyp = mz_pgrepr::Type::from(&typ.scalar_type);
+                            match mz_pgrepr::Value::decode_text(&pgtyp, field.as_bytes()) {
+                                Ok(value) => value.into_datum(&self.arena, &pgtyp),
+                                Err(_) => {
+                                    if field.is_empty() && typ.nullable {
+                                       return Datum::Null;
+                                    }
+                                  error!("could not decode typ {typ:?} for value '{field}' in ldbc output {}", csv.output);
+                                    panic!("could not decode typ {typ:?} for value '{field}' in ldbc output {}", csv.output);
+                                }
+                            }
+                        },
+                    ));
+                    // Make symmetric.
+                    if csv.output == LDBC_PERSON_KNOWS_PERSON_OUTPUT {
+                        assert!(self.pending.is_none());
+                        let mut iter = row.iter();
+                        let creationdate = iter.next().expect("must exist");
+                        let person1id = iter.next().expect("must exist");
+                        let person2id = iter.next().expect("must exist");
+                        let swapped = Row::pack_slice(&[
+                            creationdate,
+                            person2id,
+                            person1id,
+                        ]);
+                        self.pending = Some((csv.output, swapped));
+                    }
+                    Some((csv.output, row))
+                }
+                Ok(false) => None,
+                Err(_) => None,
+            })
+    }
+}
+
+impl Iterator for Context {
+    type Item = (usize, Event<Option<MzOffset>, (Row, i64)>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((output, row)) = self.pending.take() {
+            self.offset += 1;
```

Comment:
Hm, this will make the data trickle in but downloading a file looks like an atomic operation so I think this should be zero and jump to the empty antichain when everything has finished being processed

Reply:
Also, without doing this, the file changing next time we download it will definitely be non-definite

Also, then we can avoid re-downloading if the `resume_upper != [0]`